logstash-core/lib/logstash/instrument/metric_store.rb

```diff
@@ -0,0 +1,107 @@
+# encoding: utf-8
+require "concurrent"
+require "logstash/event"
+require "logstash/instrument/metric_type"
+
+module LogStash module Instrument
+  # The Metric store the data structure that make sure the data is
+  # saved in a retrievable way, this is a wrapper around multiples ConcurrentHashMap
+  # acting as a tree like structure.
+  class MetricStore
+    class ConcurrentMapExpectedError < Exception; end
+
+    def initialize
+      @store = Concurrent::Map.new
+    end
+
+    # This method use the namespace and key to search the corresponding value of
+    # the hash, if it doesn't exist it will create the appropriate namespaces
+    # path in the hash and return `new_value`
+    #
+    # @param [Array] The path where the values should be located
+    # @param [Object] The default object if the value is not found in the path
+    # @return [Object] Return the new_value of the retrieve object in the tree
+    def fetch_or_store(namespaces, key, default_value = nil)
+      fetch_or_store_namespaces(namespaces).fetch_or_store(key, block_given? ? yield(key) : default_value)
+    end
+    
+    # This method allow to retrieve values for a specific path,
+    #
+    # Also support `*` as a globbing path
+    #
+    # @param [Array] The path where values should be located
+    # @return nil if the values are not found
+    def get(*key_paths)
+      get_recursively(key_paths, @store)
+    end
+
+    # Take all the individuals `MetricType` and convert them to `Logstash::Event`
+    #
+    # @return [Array] An array of all metric transformed in `Logstash::Event`
+    def to_events
+      to_events_recursively(@store).flatten
+    end
+
+    private
+    def get_recursively(key_paths, map)
+      key_candidate = key_paths.shift
+
+      if key_paths.empty?
+        return map[key_candidate]
+      elsif (next_map = map[key_candidate]).is_a?(Concurrent::Map)
```

Comment:
I know this is personal taste, but always found this kind of lines  doing too much things at once, can we break it somehow? it makes reading the code easier, at less to me.